Sources/SwiftProtobuf/ProtobufDecoder.swift

```diff
@@ -21,564 +21,205 @@
-    let varint: UInt64
-    let unknown: UnsafeBufferPointer<UInt8>
-    let scanner: ProtobufScanner
-    var consumed = false
+public struct ProtobufDecoder: FieldDecoder {
+    // Used only by packed repeated enums; see below
+    internal var unknownOverride: Data?
 
-    init(varint: UInt64, unknown: UnsafeBufferPointer<UInt8>, scanner: ProtobufScanner) {
-        self.varint = varint
-        self.unknown = unknown
-        self.scanner = scanner
-    }
-
-    mutating func decodeSingularField<S: FieldType>(fieldType: S.Type, value: inout S.BaseType?) throws {
-        consumed = try S.setFromProtobufVarint(varint: varint, value: &value)
-    }
-
-    mutating func decodeRepeatedField<S: FieldType>(fieldType: S.Type, value: inout [S.BaseType]) throws {
-        consumed = try S.setFromProtobufVarint(varint: varint, value: &value)
-    }
+    // Current position
+    private var p : UnsafePointer<UInt8>
+    // Remaining bytes in input.
+    private var available : Int
+    // Position of start of field currently being parsed
+    private var fieldStartP : UnsafePointer<UInt8>
+    // Position of end of field currently being parsed, nil if we don't know.
+    private var fieldEndP : UnsafePointer<UInt8>?
+    // Whether or not this field has actually been used
+    private var consumed = false
+    // Wire format for last-examined field
+    private(set) var fieldWireFormat: WireFormat = .varint
+    // Field number for last-parsed field tag
+    private(set) var fieldNumber: Int = 0
+    // Collection of extension fields for this decode
+    private var extensions: ExtensionSet?
 
-    mutating func decodePackedField<S: FieldType>(fieldType: S.Type, value: inout [S.BaseType]) throws {
-        consumed = try S.setFromProtobufVarint(varint: varint, value: &value)
-    }
+    var unknownData: Data?
 
-    mutating func asProtobufUnknown(protoFieldNumber: Int) throws -> Data? {
-        return consumed ? nil : Data(buffer: unknown)
-    }
-}
+    public var complete: Bool {return available == 0}
 
-private struct FieldWireTypeFixed64: ProtobufFieldDecoder {
-    let fixed8: [UInt8]
-    let unknown: UnsafeBufferPointer<UInt8>
-    let scanner: ProtobufScanner
-    var consumed = false
 
-    init(fixed8: [UInt8], unknown: UnsafeBufferPointer<UInt8>, scanner: ProtobufScanner) {
-        self.fixed8 = fixed8
-        self.unknown = unknown
-        self.scanner = scanner
+    internal init(protobufPointer: UnsafePointer<UInt8>, count: Int, extensions: ExtensionSet? = nil) {
+        // Assuming baseAddress is not nil.
+        p = protobufPointer
+        available = count
+        fieldStartP = p
+        self.extensions = extensions
     }
 
-    mutating func decodeSingularField<S: FieldType>(fieldType: S.Type, value: inout S.BaseType?) throws {
-        try S.setFromProtobufFixed8(fixed8: fixed8, value: &value)
-        consumed = true
+    public mutating func reset() {
+        consumed = false
```

Comment:
I just removed reset() entirely as it's no longer needed.  (The switch to passing decoders via generics removed the need to access the decoder through a protocol witness.)